src/hotspot/cpu/x86/x86_64.ad

```diff
@@ -8580,6 +8580,68 @@ instruct divL_rReg(rax_RegL rax, rdx_RegL rdx, no_rax_rdx_RegL div,
   ins_pipe(ialu_reg_reg_alu0);
 %}
 
+instruct udivI_rReg(rax_RegI rax, no_rax_rdx_RegI div, rFlagsReg cr, rdx_RegI rdx)
+%{
+  match(Set rax (UDivI rax div));
+  effect(KILL rdx, KILL cr);
+
+  ins_cost(300);
+  format %{ "udivl $rax,$rax,$div\t# UDivI\n" %}
+  ins_encode %{
+    Label done;
+    Label neg_divisor_fastpath;
+    Register divisor = $div$$Register;
+    Register tmp_rdx = $rdx$$Register; 
+    Register rax = $rax$$Register;
+    __ cmpl(divisor, 0);
+    __ jccb(Assembler::less, neg_divisor_fastpath);
+    __ xorl(tmp_rdx, tmp_rdx);
+    __ divl(divisor);
+    __ jmp(done);
+    __ bind(neg_divisor_fastpath); 
+    // Fastpath for divisor < 0: 
+    // quotient = (dividend & ~(dividend - divisor)) >>> (Integer.SIZE - 1)
+    // See Hacker's Delight (2nd ed), section 9.3 which is implemented in java.lang.Long.divideUnsigned()
+    __ movl(tmp_rdx, rax);
+    __ subl(tmp_rdx, divisor);
+    __ andnl(rax, tmp_rdx, rax);
+    __ shrl(rax, 31);
+    __ bind(done);
+  %}
+  ins_pipe(ialu_reg_reg_alu0);
+%}
+
+instruct udivL_rReg(rax_RegL rax, no_rax_rdx_RegL div, rFlagsReg cr, rdx_RegL rdx)
+%{
+  match(Set rax (UDivL rax div));
+  effect(KILL rdx, KILL cr);
+
+  ins_cost(300);
+  format %{ "udivq $rax,$rax,$div\t# UDivL\n" %}
+  ins_encode %{
+    Label done;
+    Label neg_divisor_fastpath;
+    Register divisor = $div$$Register;
+    Register tmp_rdx = $rdx$$Register;;
+    Register rax = $rax$$Register;;
+    __ cmpq(divisor, 0);
+    __ jccb(Assembler::less, neg_divisor_fastpath);
+    __ xorl(tmp_rdx, tmp_rdx);
+    __ divq(divisor);
+    __ jmp(done);
+    __ bind(neg_divisor_fastpath);
+    // Fastpath for divisor < 0: 
```

Comment:
Move in macro assembly rountine.

Reply:
Sure, will move it to a macro assembly routine